drakrun/drakrun/draksetup.py

```diff
@@ -84,6 +84,50 @@ def check_root():
         return True
 
 
+@click.command(help='Cleanup the changes made by draksetup')
+def cleanup():
+    if not check_root():
+        return
+
+    install_info = InstallInfo.try_load()
+
+    if install_info is None:
+        logging.error("The cleanup has been performed")
+        return
+
+    logging.info("Ensuring that drakrun@* services are stopped...")
+    try:
+        subprocess.check_output('systemctl stop \'drakrun@*\'', shell=True, stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError:
+        raise Exception("Drakrun services not stopped")
+
+    backend = get_storage_backend(install_info)
+    vm_ids = get_all_vm_conf()
+
+    for vm_id in vm_ids:
+        vm = VirtualMachine(backend, vm_id)
+        vm.destroy()
+
+        net_enable = int(conf['drakrun'].get('net_enable', '0'))
+        out_interface = conf['drakrun'].get('out_interface', '')
+        dns_server = conf['drakrun'].get('dns_server', '')
+
+        delete_vm_network(vm_id=vm_id, net_enable=net_enable, out_interface=out_interface, dns_server=dns_server)
+        if net_enable:
+            stop_dnsmasq(vm_id=vm_id)
+
+        backend.delete_vm_volume(vm_id)
+
+        delete_vm_conf(vm_id)
+
+    if install_info.zfs_tank_name is not None:
+        backend.delete_zfs_tank()
```

Comment:
I belive this is breaking Liskov Substitution Principle. You should rather introduce some interface method like `destroy_storage` which would for qcow2 would have an empty implementation and for ZFS it would delete the tank. Then, this `destroy_storage` method should be called unconditionally here, for any storage backend.

**But** I think there is another more important issue. I think we don't explicitly create the ZFS tank in `draksetup install`, we rather expect the ZFS tank to be already present (am I right?). Thus, if we don't create it, we shouldn't also destroy it.

Speaking more generally, in every possible scenario it should be possible to call `draksetup install` -> `draksetup postinstall` -> `draksetup cleanup` -> `draksetup install` -> `draksetup postinstall` without calling any ZFS/LVM/QCOW2 related commands in between and it should work. Right now I think the second `draksetup install` would fail because `draksetup cleanup` would delete the ZFS tank but `draksetup install` would not create it again.

Reply:
> I belive this is breaking Liskov Substitution Principle. You should rather introduce some interface method like `destroy_storage` which would for qcow2 would have an empty implementation and for ZFS it would delete the tank. Then, this `destroy_storage` method should be called unconditionally here, for any storage backend.

I wanted to abstract this actually but I wasn't sure of the design. Though, I didn't know about `Liskov Substitution Principle`. Thanks for the term

> **But** I think there is another more important issue. I think we don't explicitly create the ZFS tank in `draksetup install`, we rather expect the ZFS tank to be already present (am I right?). Thus, if we don't create it, we shouldn't also destroy it.

Ah! A big blunder from my side, Apologies for that. I didn't really have zfs installed to test the `zfs` specific codebase and installing `zfs` on debian buster was causing some issues for me. I will remove this function now

> Speaking more generally, in every possible scenario it should be possible to call `draksetup install` -> `draksetup postinstall` -> `draksetup cleanup` -> `draksetup install` -> `draksetup postinstall` without calling any ZFS/LVM/QCOW2 related commands in between and it should work. Right now I think the second `draksetup install` would fail because `draksetup cleanup` would delete the ZFS tank but `draksetup install` would not create it again.

Yeah, I had that in mind but somehow missed it. Thanks for pointing it out.

Also, I wanted to ask one thing. Does the CI run `draksetup` in it? If somehow such issues occur. How do I improve my testing. Manual testing takes too much time

Reply:
> Does the CI run draksetup in it?

It does it only partially due to speed optimizations. We don't have full end to end test for draksetup at the moment.

Reply:
o_O

```
"self.zfs_tank_name"
```

i'm going to test this one, this surely looks like some mistake in the code. This needs to be corrected first.

Reply:
> o_O
> 
> ```
> "self.zfs_tank_name"
> ```
> 
> i'm going to test this one

Sure

> this surely looks like some mistake in the code. This needs to be corrected first.

Yeah, I had noticed that and fixed it in the function thinking the tanks needed to be destroyed here as the code existed

Reply:
@icedevml have you tested the code yet?

Reply:
The mistake is corrected now ✅